### PR TITLE
Implement Pattern color space and Shading, seal IColor classes

### PR DESCRIPTION
**Pull request by BobLd**
- Implement `PatternColorSpaceDetails` and `PatternColor`
- Implement `Shading` (not yet used in `ContentStreamProcessor`)
- Add color space flow shart from Pdf 1.7 documentation to be added to wiki
- Remove `Details` from `ColorSpaceDetails` property names (will add a note for breaking changes)
- Seal IColor implementations
- Use `double` instead of `decimal` in color spaces

### CODE REVIEW

File: src/UglyToad.PdfPig/Graphics/Colors/ColorSpaceDetails.cs [hunk 1]
```diff
@@ -26,20 +26,20 @@ public abstract class ColorSpaceDetails
...
+        /// The underlying type of <see cref="ColorSpace"/>, usually equal to <see cref="Type"/>
+        /// unless <see cref="ColorSpace.Indexed"/> or <see cref="ColorSpace.DeviceN"/>.
         /// </summary>
         public ColorSpace BaseType { get; protected set; }
 
         /// <summary>
         /// The number of components for the underlying color space.
         /// </summary>
-        public abstract int BaseNumberOfColorComponents { get; }
+        internal abstract int BaseNumberOfColorComponents { get; }
```
Comment (BobLd) [inline src/UglyToad.PdfPig/Graphics/Colors/ColorSpaceDetails.cs:37]: Make internal as only used in the `PngFromPdfImageFactory`

File: src/UglyToad.PdfPig/Graphics/Colors/ColorSpaceDetails.cs [hunk 2]
```diff
@@ -501,16 +502,17 @@ public sealed class DeviceNColorSpaceDetails : ColorSpaceDetails
...
             TintFunction = tintFunction;
+            BaseType = AlternateColorSpace.Type;
         }
 
         /// <inheritdoc/>
         internal override double[] Process(params double[] values)
         {
-            var evaled = TintFunction.Eval(values[0]);
-            return AlternateColorSpaceDetails.Process(evaled);
+            var evaled = TintFunction.Eval(values);
```
Comment (BobLd) [inline src/UglyToad.PdfPig/Graphics/Colors/ColorSpaceDetails.cs:514]: Fix small bug

File: src/UglyToad.PdfPig/Images/ColorSpaceDetailsByteConverter.cs [hunk 3]
```diff
@@ -53,10 +53,12 @@ private static byte[] UnpackComponents(IReadOnlyList<byte> input, int bitsPerCom
         {
             IEnumerable<byte> Unpack(byte b)
             {
+                int right = (int)Math.Pow(2, bitsPerComponent) - 1;
```
Comment (BobLd) [inline src/UglyToad.PdfPig/Images/ColorSpaceDetailsByteConverter.cs:56]: Compute outside of `for` loop. value does not change

**Review by EliotJones — APPROVED** [review]
Thanks for all the work here

**Comment by BobLd** [comment]
@EliotJones I will put will a note here for all breaking changes of my recent commits this weekend

**Comment by BobLd** [comment]
@EliotJones please find below my changes related to the color spaces and the breaking changes:

#615
- Implement `PatternColorSpaceDetails`
- Add `TilingPatternColor` and `ShadingPatternColor` that implement `PatternColor` abstract class
- Implement all `Shading` classes

**Breaking changes**
- `IColor` can now be of type `PatternColor`. This implementation will throw an error when calling `ToRGBValues()`. You might have to check for `IColor.ColorSpace != ColorSpace.Pattern` before calling the function
- Remove Details from `ColorSpaceDetails` property names:
  - `AlternateColorSpaceDetails` renamed to `AlternateColorSpace`
  - `BaseColorSpaceDetails` renamed to `BaseColorSpace`
- Seal `IColor` implementations
- Use `double` instead of `decimal` in color spaces and colors

#608
- Set rendering intent in operation

#607
- Implement Lab and DeviceN color spaces
- Refactoring `ColorSpaceDetailsByteConverter`

#595
- Use `ColorSpaceDetails` instead of ColorSpace `enum`
- Move color spaces logic inside respective `ColorSpaceDetails`
- Actually use separation color space
- ResourceStore: Store loaded name color spaces and make `namedColorSpaces` a `StackDictionary`

**Breaking changes**
- Move `IColorSpaceContext` from `IOperationContext` to `CurrentGraphicsState`
- Removed `ColorSpace` property from `IPdfImage`. Use `ColorSpaceDetails.Type` to get the `enum` value
- `IColorSpaceContext`'s `CurrentStrokingColorSpace` and `CurrentNonStrokingColorSpace` are now of type `ColorSpaceDetails` (not a `ColorSpace` `enum` anymore). Use `CurrentStrokingColorSpace.Type` or `CurrentNonStrokingColorSpace.Type` to get the `enum` value

**Pull request merged**